### ISSUE TITLE
Generic with capture returned from lambda parameter results in incorrect circularity error.

### ISSUE THREAD

**Issue by MicahZoltu** [microsoft/TypeScript#38451]
**TypeScript Version:**  `4.0.0-dev.20200509` && `3.9.0-beta` && `3.8.3`

**Search Terms:**
generic capture implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer

**Code**

```ts
declare function getApple<T>(input: T, fun: () => void): T;
const apple = getApple('apple', () => apple)
```

**Expected behavior:**
No error.

**Actual behavior:**
Error on line two:
> 'apple' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.

**Playground Link:** 
https://www.typescriptlang.org/play?ts=3.9.0-beta#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwHMQMBBABzIhAB4AVAPgAotUzkMAueWgGiTS6MAlPAC89eADccWYEK60A3ACgA9KvgByKBSqb4WALaUsYLBggBPeAAsoAZ3gZLZBNtSX9AI3BRk9hHN4YBwQR1QcDFsoSQQoJxc41AiMKGw8eChUYANHOEQQOHQQHOAsOEwreBwYA2zy8AtrFgMMRxwAd3wWcywoCCwAL0KAOmUwPHsonUoEUUJiclnGbV0QTT5hMQkZqiEgA

**Related Issues:** 
Yes, though most appear to be a correct error and people don't understand (maybe this one is too and I'm one of those people!)

**Workaround:**
Wrapping the lambda in curly braces so the captured variable is not returned resolves the issue.  Also, if the function is not generic the issue is resolved.
```ts
declare function getApple<T>(input: T, fun: () => void): T;
const apple = getApple('apple', () => {apple}) // ok
```
```ts
declare function getApple(input: string, fun: () => void): string;
const apple = getApple('apple', () => apple)
```

**Comment by MartinJohns** [comment]
const apple = getApple('apple', () => {apple}) // ok

This should probably be `() => ({apple})` to be an equivalent example. Here the function returns an object, in your example it returns nothing.

**Comment by MicahZoltu** [comment]
@MartinJohns Returning void was specifically the fix I was trying to illustrate.  If the function doesn't *return* the captured value, then the type is inferred correctly.  If the function *does* return the captured value, it infers incorrectly.

Using your suggested code, returning an object with the captured value as a property results in inference failure again.
```ts
declare function getApple<T>(input: T, fun: () => void): T;
const apple = getApple('apple', () => ({apple})) // error
```

**Comment by RyanCavanaugh** [comment]
We issue errors on circularity, and can *sometimes* avoid doing so. The fact that we can *sometimes* avoid hitting a circularity error when one is present should be taken as a carve-out; if you have a circularity, you *may* receive a circularity error even if you don't immediately see the path the checker to get into that state.

**Comment by MicahZoltu** [comment]
@RyanCavanaugh Can you help me understand how this is circular according to the compiler?  The passed in function is not part of `T` at all so it seems like the compiler should be ignoring it entirely when inferring `T`.

In the provided example `T` (and thus the type of `apple`) should be the type of the first parameter of the function, the second parameter plays no role in determining `T`.

**Comment by MicahZoltu** [comment]
With some help in TS Discord I *think* I understand the problem here.  TypeScript is trying to figure out what the type of each of the parameters to `getApple` are to see if they match the `getApple` signature *prior* to figuring out what the type of `apple` is.  So first the compiler figures out that the type of the first parameter is `'apple'` and then it tries to figure out what the type of the second parameter is but runs into a problem because it doesn't yet know the type of `apple`.  At this point it throws an error, because it has given up on type checking the function call at this point.

As a human, I am starting with the signature types and then checking to see if the provided parameters match, but the compiler is first trying to fully resolve the provided parameter types and then comparing them to see if they match.  This means that the compiler doesn't short-circuit with "we don't actually care what the return type of `() => apple` is because the function's signature calls for `=> void`" like the human does.
